src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/CSupport.java

```diff
@@ -49,6 +50,163 @@ public static ForeignLinker getSystemLinker() {
+         *
+         * @param layout the layout of the value
+         * @return the value read as an {@code int}
+         */
+        int readInt(MemoryLayout layout);
+
+        /**
+         * Reads a value into a {@code long}
+         *
+         * @param layout the layout of the value
+         * @return the value read as an {@code long}
+         */
+        long readLong(MemoryLayout layout);
+
+        /**
+         * Reads a value into a {@code double}
+         *
+         * @param layout the layout of the value
+         * @return the value read as an {@code double}
+         */
+        double readDouble(MemoryLayout layout);
+
+        /**
+         * Reads a value into a {@code MemoryAddress}
+         *
+         * @param layout the layout of the value
+         * @return the value read as an {@code MemoryAddress}
+         */
+        MemoryAddress readPointer(MemoryLayout layout);
+
+        /**
+         * Reads a value into a {@code MemorySegment}
+         *
+         * @param layout the layout of the value
+         * @return the value read as an {@code MemorySegment}
+         */
+        MemorySegment readStructOrUnion(MemoryLayout layout);
+
+        /**
+         * Skips a number of va arguments with the given memory layouts.
+         *
+         * @param layouts the layout of the value
+         */
+        void skip(MemoryLayout...layouts);
+
+        /**
+         * A predicate used to check if this va list is alive,
+         * or in other words; if {@code close()} has been called on this
+         * va list.
+         *
+         * @return true if this va list is still alive.
+         * @see #close()
+         */
+        boolean isAlive();
+
+        /**
+         * Closes this va list, releasing any resources it was using.
+         *
+         * @see #isAlive()
+         */
+        void close();
+
+        /**
+         * Copies this va list.
+         *
+         * @return a copy of this va list.
+         */
+        VaList copy();
+
+        /**
+         * Returns the underlying memory address of this va list.
+         *
+         * @return the address
+         */
+        MemoryAddress toAddress();
+
+        /**
+         * Constructs a {@code VaList} out of the memory address of a va_list.
+         *
+         * @param ma the memory address
+         * @return the new {@code VaList}.
+         */
+        static VaList ofAddress(MemoryAddress ma) {
+            return SharedUtils.newVaListOfAddress(ma);
+        }
+
+        /**
+         * A builder interface used to construct a va list.
+         */
+        interface Builder {
+
+            /**
+             * Adds a native value represented as an {@code int} to the va list.
+             *
+             * @param layout the native layout of the value.
+             * @param value the value, represented as an {@code int}.
+             * @return this builder.
+             */
+            Builder intArg(MemoryLayout layout, int value);
```

Comment:
Dual issue here - `int` again stands for Java `int`. Perthaps using `asInt` in the method name could help clarify.